### ISSUE TITLE
no-useless-assertion: false positive on non-null-assertion

### ISSUE THREAD

**Issue by AlCalzone** [fimbullinter/wotan#719]
Repro:

```ts
interface ImportRange {
	index: number;
	end: number;
	module: string;
	symbol: string;
	import: string;
	options: {
		comments?: boolean;
		jsdoc?: boolean;
	};
}

const importRegex = /(?<import><!-- #import (?<symbol>.*?) from "(?<module>.*?)"(?: with (?<options>[\w\-, ]*?))? -->)(?:[\s\r\n]*```ts[\r\n]*(?<source>(.|\n)*?)```)?/g;

export function findImportRanges(docFile: string): ImportRange[] {
	const matches = [...docFile.matchAll(importRegex)];
	return matches.map((match) => ({
		index: match.index!,
		// eslint-disable-next-line @typescript-eslint/restrict-plus-operands
		end: match.index! + match[0].length,
		module: match.groups!.module,
		symbol: match.groups!.symbol,
		import: match.groups!.import,
		options: {
			comments: !!match.groups!.options?.includes("comments"),
			jsdoc: !match.groups!.options?.includes("no-jsdoc"),
		},
	}));
}
```

The rule `no-useless-assertion` complains about every non-null assertion in `findImportRanges`. If I remove one, TypeScript complains that `undefined` is not assignable to the target type.

**Comment by ajafff** [comment]
I cannot reproduce that locally with `typescript@latest` and `typescript@next` and only `no-useless-assertion` enabled.
What version of TypeScript are you using?
Are the findings from the language-service plugin? If so, does it reproduce using the CLI?

I guess something is messing with the types. Maybe caused by another rule, or maybe typescript-eslint if it runs as a language-service plugin too.

**Comment by AlCalzone** [comment]
It seems to reproduce only on the CLI. To repro:
1. checkout https://github.com/zwave-js/node-zwave-js/commit/2d08d5b8d7c968afcabb33a9af1cfefa48a0fabb
2. `npm ci`
3. open `maintenance\generateTypedDocs.ts`, remove the comment in line 5
4. `npm run lint:wotan`

**Comment by ajafff** [comment]
This is caused by `Property 'matchAll' does not exist on type 'string'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2020' or later.`

The tsconfig you use for linting has `target: es2018` and no `lib`. The config you use for building this directory has `lib: es2020`. Therefore the compiler and the linter disagree on types.

**Comment by AlCalzone** [comment]
Ew - thanks for the investigation. I'll see what I can do about that.
